src/shogun/machine/gp/SingleFITCLaplacianBase.cpp

```diff
@@ -0,0 +1,409 @@
+}
+
+float64_t CSingleFITCLaplacianBase::get_derivative_related_cov_helper(
+	SGMatrix<float64_t> dKuui, SGVector<float64_t> v, SGMatrix<float64_t> R)
+{
+	Map<VectorXd> eigen_w(m_w.vector, m_w.vlen);
+	Map<MatrixXd> eigen_W(m_Rvdd.matrix, m_Rvdd.num_rows, m_Rvdd.num_cols);
+	Map<MatrixXd> eigen_B(m_B.matrix, m_B.num_rows, m_B.num_cols);
+
+	Map<MatrixXd> eigen_dKuui(dKuui.matrix, dKuui.num_rows, dKuui.num_cols);
+	Map<VectorXd> eigen_v(v.vector, v.vlen);
+	Map<MatrixXd> eigen_R(R.matrix, R.num_rows, R.num_cols);
+
+	//-al'*(v.*al)-sum(W.*W,1)*v = v'*(-sum(W.*W,1)'-(al.*al))
+	SGVector<float64_t> di=get_derivative_related_cov_diagonal();
+	Map<VectorXd> eigen_di(di.vector, di.vlen);
+
+	//(w'*dKuui*w -al'*(v.*al)- sum(W.*W,1)*v - sum(sum((R*W').*BWt)))/2; 
+	float64_t result=(eigen_w.dot(eigen_dKuui*eigen_w)+eigen_v.dot(eigen_di)-
+			(eigen_R*eigen_W.adjoint()).cwiseProduct(eigen_B*eigen_W.adjoint()).sum())/2.0;
+
+	return result;
+}
+
+float64_t CSingleFITCLaplacianBase::get_derivative_related_cov(SGVector<float64_t> ddiagKi,
+	SGMatrix<float64_t> dKuui, SGMatrix<float64_t> dKui)
+{
+	Map<MatrixXd> eigen_B(m_B.matrix, m_B.num_rows, m_B.num_cols);
+	Map<VectorXd> eigen_ddiagKi(ddiagKi.vector, ddiagKi.vlen);
+	Map<MatrixXd> eigen_dKuui(dKuui.matrix, dKuui.num_rows, dKuui.num_cols);
+	Map<MatrixXd> eigen_dKui(dKui.matrix, dKui.num_rows, dKui.num_cols);
+
+	// compute R=2*dKui-dKuui*B
+	SGMatrix<float64_t> R(dKui.num_rows, dKui.num_cols);
+	Map<MatrixXd> eigen_R(R.matrix, R.num_rows, R.num_cols);
+	eigen_R=2*eigen_dKui-eigen_dKuui*eigen_B;
+
+	// compute v=ddiagKi-sum(R.*B, 1)'
+	SGVector<float64_t> v(ddiagKi.vlen);
+	Map<VectorXd> eigen_v(v.vector, v.vlen);
+	eigen_v=eigen_ddiagKi-eigen_R.cwiseProduct(eigen_B).colwise().sum().adjoint();
+
+	return get_derivative_related_cov(ddiagKi, dKuui, dKui, v, R);
+}
+
+float64_t CSingleFITCLaplacianBase::get_derivative_related_cov(SGVector<float64_t> ddiagKi,
+	SGMatrix<float64_t> dKuui, SGMatrix<float64_t> dKui,
+	SGVector<float64_t> v, SGMatrix<float64_t> R)
+{
+	Map<VectorXd> eigen_t(m_t.vector, m_t.vlen);
+	Map<VectorXd> eigen_al(m_al.vector, m_al.vlen);
+	Map<VectorXd> eigen_w(m_w.vector, m_w.vlen);
+	Map<MatrixXd> eigen_W(m_Rvdd.matrix, m_Rvdd.num_rows, m_Rvdd.num_cols);
+	Map<VectorXd> eigen_ddiagKi(ddiagKi.vector, ddiagKi.vlen);
+	Map<MatrixXd> eigen_dKuui(dKuui.matrix, dKuui.num_rows, dKuui.num_cols);
+	Map<MatrixXd> eigen_dKui(dKui.matrix, dKui.num_rows, dKui.num_cols);
+
+	//(w'*dKuui*w -al'*(v.*al)- sum(W.*W,1)*v - sum(sum((R*W').*BWt)))/2; 
+	float64_t result=get_derivative_related_cov_helper(dKuui, v, R);
+
+	// compute dnlZ=(ddiagKi'*(1./g_sn2)+w'*(dKuui*w-2*(dKui*al))-al'*(v.*al)-
+	// sum(W.*W,1)*v- sum(sum((R*W').*(B*W'))))/2;
+	result+=(eigen_ddiagKi.dot(eigen_t))/2.0-
+			eigen_w.dot((eigen_dKui*eigen_al));
+	return result;
+
+}
+
+SGVector<float64_t> CSingleFITCLaplacianBase::get_derivative_wrt_inference_method(
+		const TParameter* param)
+{
+	REQUIRE(!(strcmp(param->m_name, "scale")
```

Comment:
Put a REQUIRE(param, "param not set\n") here to avoid segfaults in the next check